File: js-old/src/util/tx.js
// Copyright 2015-2017 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

import BigNumber from 'bignumber.js';

import WalletsUtils from '~/util/wallets';

/**
 * The sender is by default (when the UI loads) the
 * default dapp address. It can then be modified when
 * sending transactions....
 */
let currentSender = '';
let hasCurrentSenderChanged = false;

export function getSender () {
  currentSender;
}

export function loadSender (api) {
  // If the current sender has been changed
  // then don't bother checking changes of the
  // default sender
  if (hasCurrentSenderChanged) {
    return Promise.resolve(currentSender);
  }

  return api.parity.getNewDappsDefaultAddress()
    .then((defaultAccount) => {
      currentSender = defaultAccount;
      return defaultAccount;
    });
}

export function setSender (sender) {
  currentSender = sender;
  hasCurrentSenderChanged = true;
}

export function trackRequest (api, options, statusCallback) {
  const { requestId, transactionHash } = options;
  const txHashPromise = transactionHash
    ? Promise.resolve(transactionHash)
    : api.pollMethod('parity_checkRequest', requestId);

  return txHashPromise
    .then((transactionHash) => {
      statusCallback(null, { transactionHash });
      return api.pollMethod('eth_getTransactionReceipt', transactionHash, isValidReceipt);
    })
    .then((transactionReceipt) => {
      statusCallback(null, { transactionReceipt });
    })
    .catch((error) => {
      statusCallback(error);
    });
}

const isValidReceipt = (receipt) => {
  return receipt && receipt.blockNumber && receipt.blockNumber.gt(0);
};

export function getTxOptions (api, func, _options, values = []) {
  const options = { ..._options };
  const address = options.from;

  if (func && func.contract) {
    options.to = options.to || func.contract.address;
  }

  const promise = (!address)
    ? Promise.resolve(false)
    : WalletsUtils.isWallet(api, address);

  return promise
    .then((isWallet) => {
      if (!isWallet) {
        return { func, options, values };
      }

      if (func && func.contract) {
        options.data = func.contract.getCallData(func, options, values);
      }

      if (!options.to) {
        return { func, options, values };
      }

      return WalletsUtils
        .getCallArgs(api, options, values)
        .then((callArgs) => {
          if (!callArgs) {
            return { func, options, values };
          }

          return callArgs;
        });
    })
    .then(({ func, options, values }) => {
      if (func) {
        options.data = func.contract.getCallData(func, options, values);
      }

      if (!options.value) {
        options.value = new BigNumber(0);
      }

      return options;
    });
}

export function estimateGas (_func, _options, _values = []) {
  const { api } = _func.contract;

  return getTxOptions(api, _func, _options, _values)
    .then((options) => {
      return api.eth.estimateGas(options);
    });
}

export function postTransaction (_func, _options, _values = []) {
  const { api } = _func.contract;

  return getTxOptions(api, _func, _options, _values)
    .then((options) => {
      return api.parity.postTransaction(options);
    });
}

export function deployEstimateGas (contract, _options, values) {
  const options = { ..._options };
  const { api } = contract;
  const address = options.from;

  return WalletsUtils
    .isWallet(api, address)
    .then((isWallet) => {
      if (!isWallet) {
        return contract.deployEstimateGas(options, values);
      }

      return WalletsUtils
        .getDeployArgs(contract, options, values)
        .then((callArgs) => {
          const { func, options, values } = callArgs;

          return func.estimateGas(options, values);
        })
        .then((gasEst) => {
          return [gasEst, gasEst.mul(1.05)];
        });
    });
}

export function deploy (contract, options, values, skipGasEstimate = false) {
  const { api } = contract;
  const address = options.from;

  const gasEstPromise = skipGasEstimate
    ? Promise.resolve(null)
    : deployEstimateGas(contract, options, values).then(([gasEst, gas]) => gas);

  return gasEstPromise
    .then((gas) => {
      if (gas) {
        options.gas = gas.toFixed(0);
      }

      return WalletsUtils.isWallet(api, address);
    })
    .then((isWallet) => {
      if (!isWallet) {
        const encodedOptions = contract._encodeOptions(contract.constructors[0], options, values);

        return api.parity.postTransaction(encodedOptions);
      }

      return WalletsUtils.getDeployArgs(contract, options, values)
        .then((callArgs) => {
          const { func, options, values } = callArgs;

          return func._postTransaction(options, values);
        });
    });
}

export function parseTransactionReceipt (api, options, receipt) {
  if (receipt.gasUsed.eq(options.gas)) {
    const error = new Error(`Contract not deployed, gasUsed == ${options.gas.toFixed(0)}`);

    return Promise.reject(error);
  }

  // If regular contract creation, only validate the contract
  if (receipt.contractAddress) {
    return validateContract(api, receipt.contractAddress);
  }

  // Otherwise, needs to check for a contract deployment
  // from a multisig wallet
  const walletResult = WalletsUtils.parseTransactionLogs(api, options, receipt.logs || []);

  if (!walletResult) {
    const error = new Error('Something went wrong in the contract deployment...');

    return Promise.reject(error);
  }

  if (walletResult.pending) {
    return Promise.resolve(null);
  }

  return validateContract(api, walletResult.contractAddress);
}

function validateContract (api, contractAddress) {
  return api.eth
    .getCode(contractAddress)
    .then((code) => {
      if (code === '0x') {
        throw new Error('Contract not deployed, getCode returned 0x');
      }

      return contractAddress;
    });
}

export function patchApi (api) {
  api.patch = {
    ...api.patch,
    contract: patchContract
  };
}

export function patchContract (contract) {
  contract._functions.forEach((func) => {
    if (!func.constant) {
      func._postTransaction = func.postTransaction;
      func._estimateGas = func.estimateGas;

      func.postTransaction = postTransaction.bind(contract, func);
      func.estimateGas = estimateGas.bind(contract, func);
    }
  });
}

export function checkIfTxFailed (api, tx, gasSent) {
  return api.pollMethod('eth_getTransactionReceipt', tx)
  .then((receipt) => {
    // TODO: Right now, there's no way to tell wether the EVM code crashed.
    // Because you usually send a bit more gas than estimated (to make sure
    // it gets mined quickly), we transaction probably failed if all the gas
    // has been used up.
    return receipt.gasUsed.eq(gasSent);
  });
}

export function waitForConfirmations (api, tx, confirmations) {
  return new Promise((resolve, reject) => {
    api.pollMethod('eth_getTransactionReceipt', tx, isValidReceipt)
    .then((receipt) => {
      let subscription;

      api.subscribe('eth_blockNumber', (err, block) => {
        if (err) {
          reject(err);
        } else if (block.minus(confirmations - 1).gte(receipt.blockNumber)) {
          if (subscription) {
            api.unsubscribe(subscription);
          }
          resolve();
        }
      })
      .then((_subscription) => {
        subscription = _subscription;
      })
      .catch(reject);
    });
  });
}

File: miner/src/pool/tests/tx.rs
// Copyright 2015-2018 Parity Technologies (UK) Ltd.
// This file is part of Parity.

// Parity is free software: you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation, either version 3 of the License, or
// (at your option) any later version.

// Parity is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.

// You should have received a copy of the GNU General Public License
// along with Parity.  If not, see <http://www.gnu.org/licenses/>.

use ethereum_types::{U256, H256};
use ethkey::{Random, Generator};
use rustc_hex::FromHex;
use transaction::{self, Transaction, SignedTransaction, UnverifiedTransaction};

use pool::{verifier, VerifiedTransaction};

#[derive(Clone)]
pub struct Tx {
	pub nonce: u64,
	pub gas: u64,
	pub gas_price: u64,
	pub multiplier: u64,
}

impl Default for Tx {
	fn default() -> Self {
		Tx {
			nonce: 123,
			gas: 21_000,
			gas_price: 1,
			multiplier: 0,
		}
	}
}

impl Tx {
	pub fn gas_price(gas_price: u64) -> Self {
		Tx {
			gas_price,
			..Default::default()
		}
	}

	pub fn signed(self) -> SignedTransaction {
		let keypair = Random.generate().unwrap();
		self.unsigned().sign(keypair.secret(), None)
	}

	pub fn signed_pair(self) -> (SignedTransaction, SignedTransaction) {
		let (tx1, tx2, _) = self.signed_triple();
		(tx1, tx2)
	}

	pub fn signed_triple(mut self) -> (SignedTransaction, SignedTransaction, SignedTransaction) {
		let keypair = Random.generate().unwrap();
		let tx1 = self.clone().unsigned().sign(keypair.secret(), None);
		self.nonce += 1;
		let tx2 = self.clone().unsigned().sign(keypair.secret(), None);
		self.nonce += 1;
		let tx3 = self.unsigned().sign(keypair.secret(), None);

		(tx1, tx2, tx3)
	}

	/// Consecutive transactions where the gas_price is decided by the predicate
	/// P(i, nonce)
	pub fn signed_consecutive<P>(mut self, amount: usize, fun: P) -> Vec<SignedTransaction>
		where P: Fn(usize, u64) -> u64
	{
		let keypair = Random.generate().unwrap();
		(0..amount).map(|i| {
			self.gas_price = fun(i, self.nonce);
			let tx = self.clone().unsigned().sign(keypair.secret(), None);
			self.nonce += 1;
			tx
		}).collect::<Vec<SignedTransaction>>()
	}

	pub fn signed_replacement(mut self) -> (SignedTransaction, SignedTransaction) {
		let keypair = Random.generate().unwrap();
		let tx1 = self.clone().unsigned().sign(keypair.secret(), None);
		self.gas_price += 1;
		let tx2 = self.unsigned().sign(keypair.secret(), None);

		(tx1, tx2)
	}

	pub fn unsigned(self) -> Transaction {
		Transaction {
			action: transaction::Action::Create,
			value: U256::from(100),
			data: "3331600055".from_hex().unwrap(),
			gas: self.gas.into(),
			gas_price: self.gas_price.into(),
			nonce: self.nonce.into()
		}
	}

	pub fn big_one(self) -> SignedTransaction {
		let keypair = Random.generate().unwrap();
		let tx = Transaction {
			action: transaction::Action::Create,
			value: U256::from(100),
			data: include_str!("../res/big_transaction.data").from_hex().unwrap(),
			gas: self.gas.into(),
			gas_price: self.gas_price.into(),
			nonce: self.nonce.into()
		};
		tx.sign(keypair.secret(), None)
	}
}

#[derive(Default)]
pub struct TxBuilder {
	pub nonce: Option<u64>,
	pub gas: Option<u64>,
	pub gas_price: Option<u64>,
	pub multiplier: Option<u64>,
}

impl TxBuilder {
	pub fn build(&mut self) -> Tx {
		let mut tx = Tx::default();
		if let Some(n) = self.nonce {
			tx.nonce = n;
		}
		if let Some(g) = self.gas {
			tx.gas = g;
		}
		if let Some(gp) = self.gas_price {
			tx.gas_price = gp;
		}
		if let Some(multiplier) = self.multiplier {
			tx.multiplier = multiplier;
		}
		tx
	}

	pub fn nonce(&mut self, nonce: u64) -> &mut Self {
		self.nonce = Some(nonce);
		self
	}

	pub fn gas(&mut self, gas: u64) -> &mut Self {
		self.gas = Some(gas);
		self
	}
	
	pub fn gas_price(&mut self, gas_price: u64) -> &mut Self {
		self.gas_price = Some(gas_price);
		self
	}

	pub fn multiplier(&mut self, multiplier: u64) -> &mut Self {
		self.multiplier = Some(multiplier);
		self
	}
}

pub trait TxExt: Sized {
	type Out;
	type Verified;
	type Hash;

	fn hash(&self) -> Self::Hash;

	fn local(self) -> Self::Out;

	fn retracted(self) -> Self::Out;

	fn unverified(self) -> Self::Out;

	fn verified(self) -> Self::Verified;
}

impl<A, B, O, V, H> TxExt for (A, B) where
	A: TxExt<Out=O, Verified=V, Hash=H>,
	B: TxExt<Out=O, Verified=V, Hash=H>,
{
	type Out = (O, O);
	type Verified = (V, V);
	type Hash = (H, H);

	fn hash(&self) -> Self::Hash { (self.0.hash(), self.1.hash()) }
	fn local(self) -> Self::Out { (self.0.local(), self.1.local()) }
	fn retracted(self) -> Self::Out { (self.0.retracted(), self.1.retracted()) }
	fn unverified(self) -> Self::Out { (self.0.unverified(), self.1.unverified()) }
	fn verified(self) -> Self::Verified { (self.0.verified(), self.1.verified()) }
}

impl TxExt for SignedTransaction {
	type Out = verifier::Transaction;
	type Verified = VerifiedTransaction;
	type Hash = H256;

	fn hash(&self) -> Self::Hash {
		UnverifiedTransaction::hash(self)
	}

	fn local(self) -> Self::Out {
		verifier::Transaction::Local(self.into())
	}

	fn retracted(self) -> Self::Out {
		verifier::Transaction::Retracted(self.into())
	}

	fn unverified(self) -> Self::Out {
		verifier::Transaction::Unverified(self.into())
	}

	fn verified(self) -> Self::Verified {
		VerifiedTransaction::from_pending_block_transaction(self)
	}
}

impl TxExt for Vec<SignedTransaction> {
	type Out = Vec<verifier::Transaction>;
	type Verified = Vec<VerifiedTransaction>;
	type Hash = Vec<H256>;

	fn hash(&self) -> Self::Hash {
		self.iter().map(|tx| tx.hash()).collect()
	}

	fn local(self) -> Self::Out {
		self.into_iter().map(Into::into).map(verifier::Transaction::Local).collect()
	}

	fn retracted(self) -> Self::Out {
		self.into_iter().map(Into::into).map(verifier::Transaction::Retracted).collect()
	}

	fn unverified(self) -> Self::Out {
		self.into_iter().map(Into::into).map(verifier::Transaction::Unverified).collect()
	}

	fn verified(self) -> Self::Verified {
		self.into_iter().map(VerifiedTransaction::from_pending_block_transaction).collect()
	}
}

pub trait PairExt {
	type Type;

	fn into_vec(self) -> Vec<Self::Type>;
}

impl<A> PairExt for (A, A) {
	type Type = A;
	fn into_vec(self) -> Vec<A> {
		vec![self.0, self.1]
	}
}
